File: tools/ck_server.py
#!/usr/bin/env python3
"""
C-Kernel-Engine Streaming Inference Server

HTTP server with Server-Sent Events (SSE) for streaming token generation.
Compatible with OpenAI-style API endpoints.

Usage:
    python tools/ck_server.py --model-dir ~/.cache/huggingface/hub/SmolLM-135M --port 8080

API Endpoints:
    POST /v1/completions     - Generate completion (streaming or non-streaming)
    POST /v1/chat/completions - Chat completion (streaming or non-streaming)
    GET  /v1/models          - List available models
    GET  /health             - Health check
"""

import argparse
import json
import os
import subprocess
import sys
import time
import threading
import queue
from http.server import HTTPServer, BaseHTTPRequestHandler
from pathlib import Path
from urllib.parse import urlparse, parse_qs
import shutil

import numpy as np

# Disable torch compile overhead
os.environ.setdefault("TORCH_DISABLE_DYNAMO", "1")
os.environ.setdefault("TORCHINDUCTOR_DISABLE", "1")
os.environ.setdefault("TORCH_COMPILE_DISABLE", "1")

from transformers import AutoTokenizer


VERSION = "0.1.0"

BANNER = r"""
   ____      _  __                    _   ____
  / ___|    | |/ /___ _ __ _ __   ___| | / ___|  ___ _ ____   _____ _ __
 | |   _____| ' // _ \ '__| '_ \ / _ \ | \___ \ / _ \ '__\ \ / / _ \ '__|
 | |__|_____| . \  __/ |  | | | |  __/ |  ___) |  __/ |   \ V /  __/ |
  \____|    |_|\_\___|_|  |_| |_|\___|_| |____/ \___|_|    \_/ \___|_|

  C-Kernel-Engine Inference Server v{version}
  By Anthony Shivakumar

  Streaming API Server for C-based Transformer Inference
  ============================================================
"""


def load_config(model_dir):
    """Load model config from HuggingFace directory."""
    config_path = os.path.join(model_dir, "config.json")
    with open(config_path, "r", encoding="utf-8") as f:
        cfg = json.load(f)
    if "text_config" in cfg:
        cfg = cfg["text_config"]
    return cfg


def detect_cc():
    """Detect best available C compiler."""
    if shutil.which("icx"):
        return "icx"
    if shutil.which("icc"):
        return "icc"
    return "gcc"


def detect_avx_flags():
    """Detect AVX capabilities."""
    try:
        with open("/proc/cpuinfo", "r", encoding="utf-8") as f:
            for line in f:
                if line.startswith("flags"):
                    flags = set(line.split(":")[1].strip().split())
                    if "avx512f" in flags:
                        return ["-mavx512f"]
                    if "avx2" in flags:
                        return ["-mavx2"]
                    if "avx" in flags:
                        return ["-mavx"]
                    break
    except OSError:
        pass
    return []


def openmp_flag(cc):
    """Get OpenMP flag for compiler."""
    cc_lower = cc.lower()
    if "icx" in cc_lower or "icc" in cc_lower:
        return "-qopenmp"
    return "-fopenmp"


def sample_token(logits, temperature=0.7, top_k=40, top_p=0.9):
    """Sample next token from logits."""
    if temperature <= 0:
        return int(np.argmax(logits))

    logits = logits / temperature

    if top_k > 0:
        indices = np.argpartition(-logits, min(top_k, len(logits) - 1))[:top_k]
        top_logits = logits[indices]
        probs = np.exp(top_logits - np.max(top_logits))
        probs = probs / probs.sum()

        if top_p < 1.0:
            sorted_idx = np.argsort(-probs)
            cumsum = np.cumsum(probs[sorted_idx])
            cutoff = np.searchsorted(cumsum, top_p) + 1
            sorted_idx = sorted_idx[:cutoff]
            probs_filtered = probs[sorted_idx]
            probs_filtered = probs_filtered / probs_filtered.sum()
            chosen = np.random.choice(sorted_idx, p=probs_filtered)
            return int(indices[chosen])
        else:
            chosen = np.random.choice(len(probs), p=probs)
            return int(indices[chosen])
    else:
        probs = np.exp(logits - np.max(logits))
        probs = probs / probs.sum()
        return int(np.random.choice(len(probs), p=probs))


class CKernelModel:
    """Wrapper for C-Kernel-Engine model inference."""

    def __init__(self, model_dir, build_dir, context_len=512):
        self.model_dir = model_dir
        self.build_dir = build_dir
        self.context_len = context_len
        self.lock = threading.Lock()

        self.project_root = str(Path(__file__).parent.parent)

        # Load tokenizer
        self.tokenizer = AutoTokenizer.from_pretrained(model_dir, use_fast=True)

        # Load config
        self.cfg = load_config(model_dir)
        self.vocab_size = self.cfg.get("vocab_size", 32000)
        self.model_name = os.path.basename(model_dir)

        # Override context
        self.cfg["max_position_embeddings"] = context_len
        self.cfg["context_window"] = context_len

        # Write runtime config
        os.makedirs(build_dir, exist_ok=True)
        self.config_path = os.path.join(build_dir, "server.config.json")
        with open(self.config_path, "w", encoding="utf-8") as f:
            json.dump(self.cfg, f, indent=2)

        # Build model
        self._build_model()

        # Prepare weights
        self._prepare_weights()

    def _build_model(self):
        """Build the model binary."""
        gen_c = os.path.join(self.build_dir, "server_model.c")
        kernel_manifest = gen_c + ".kernels"
        self.model_bin = os.path.join(self.build_dir, "server_model")

        ir_demo = os.path.join(self.project_root, "build", "ck_ir_demo")
        if not os.path.exists(ir_demo):
            subprocess.check_call(["make", "build/ck_ir_demo"], cwd=self.project_root)

        subprocess.check_call([ir_demo, self.config_path, "--emit", gen_c], cwd=self.project_root)

        with open(kernel_manifest, "r", encoding="utf-8") as f:
            kernels = f.read().split()

        cc = os.environ.get("CC", detect_cc())
        cflags = ["-O3", "-fPIC", openmp_flag(cc), "-Wall"] + detect_avx_flags() + ["-Iinclude"]

        cmd = [cc] + cflags + [gen_c] + kernels + ["-o", self.model_bin, "-lm"]
        subprocess.check_call(cmd, cwd=self.project_root)

    def _prepare_weights(self):
        """Convert weights to bump format."""
        self.weights_bin = os.path.join(self.build_dir, "server_weights.bin")

        subprocess.check_call([
            sys.executable,
            os.path.join(self.project_root, "scripts", "convert_hf_to_bump.py"),
            "--checkpoint", self.model_dir,
            "--output", self.weights_bin,
            "--config", self.config_path,
            "--context", str(self.context_len),
        ], cwd=self.project_root)

    def encode(self, text):
        """Encode text to tokens."""
        return self.tokenizer.encode(text, add_special_tokens=False)

    def decode(self, token_ids):
        """Decode tokens to text."""
        return self.tokenizer.decode(token_ids, skip_special_tokens=True)

    def forward(self, token_ids):
        """Run forward pass, return logits for last position."""
        with self.lock:
            T = len(token_ids)
            if T > self.context_len:
                token_ids = token_ids[-self.context_len:]
                T = self.context_len

            padded = list(token_ids)
            if len(padded) < self.context_len:
                pad_id = self.tokenizer.eos_token_id or 0
                padded.extend([pad_id] * (self.context_len - len(padded)))

            tokens_bin = os.path.join(self.build_dir, f"tokens_{threading.current_thread().ident}.bin")
            logits_bin = os.path.join(self.build_dir, f"logits_{threading.current_thread().ident}.bin")

            np.array(padded, dtype=np.int32).tofile(tokens_bin)

            cmd = [
                self.model_bin,
                "--model-weights", self.weights_bin,
                "--tokens", tokens_bin,
                "--out-logits", logits_bin,
                "--ctx", str(self.context_len),
            ]

            subprocess.check_call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

            logits = np.fromfile(logits_bin, dtype=np.float32)
            logits = logits.reshape(self.context_len, self.vocab_size)

            # Cleanup temp files
            try:
                os.remove(tokens_bin)
                os.remove(logits_bin)
            except OSError:
                pass

            return logits[T - 1]

    def generate_stream(self, prompt, max_tokens=128, temperature=0.7, top_k=40, top_p=0.9, stop=None):
        """Generate tokens as a stream."""
        token_ids = self.encode(prompt)

        for i in range(max_tokens):
            logits = self.forward(token_ids)
            next_token = sample_token(logits, temperature, top_k, top_p)

            if next_token == self.tokenizer.eos_token_id:
                break

            token_ids.append(next_token)
            token_text = self.decode([next_token])

            # Check stop sequences
            if stop:
                full_text = self.decode(token_ids[len(self.encode(prompt)):])
                for s in stop:
                    if s in full_text:
                        return

            yield {
                "token": next_token,
                "text": token_text,
                "index": i,
            }

    def generate(self, prompt, max_tokens=128, temperature=0.7, top_k=40, top_p=0.9, stop=None):
        """Generate complete response."""
        tokens = []
        for chunk in self.generate_stream(prompt, max_tokens, temperature, top_k, top_p, stop):
            tokens.append(chunk["token"])
        return self.decode(tokens)


class CKernelHandler(BaseHTTPRequestHandler):
    """HTTP request handler for the inference server."""

    model = None
    server_start_time = None

    def log_message(self, format, *args):
        """Custom logging format."""
        timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{timestamp}] {self.address_string()} - {format % args}")

    def send_json(self, data, status=200):
        """Send JSON response."""
        body = json.dumps(data).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", len(body))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(body)

    def send_sse(self, data):
        """Send Server-Sent Event."""
        if isinstance(data, dict):
            data = json.dumps(data)
        self.wfile.write(f"data: {data}\n\n".encode("utf-8"))
        self.wfile.flush()

    def do_OPTIONS(self):
        """Handle CORS preflight."""
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")
        self.end_headers()

    def do_GET(self):
        """Handle GET requests."""
        path = urlparse(self.path).path

        if path == "/health":
            self.send_json({
                "status": "ok",
                "version": VERSION,
                "uptime": time.time() - self.server_start_time,
            })

        elif path == "/v1/models":
            self.send_json({
                "object": "list",
                "data": [{
                    "id": self.model.model_name,
                    "object": "model",
                    "owned_by": "c-kernel-engine",
                    "permission": [],
                }]
            })

        else:
            self.send_json({"error": "Not found"}, 404)

    def do_POST(self):
        """Handle POST requests."""
        path = urlparse(self.path).path

        # Read body
        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length).decode("utf-8")

        try:
            data = json.loads(body) if body else {}
        except json.JSONDecodeError:
            self.send_json({"error": "Invalid JSON"}, 400)
            return

        if path == "/v1/completions":
            self.handle_completions(data)
        elif path == "/v1/chat/completions":
            self.handle_chat_completions(data)
        else:
            self.send_json({"error": "Not found"}, 404)

    def handle_completions(self, data):
        """Handle /v1/completions endpoint."""
        prompt = data.get("prompt", "")
        max_tokens = data.get("max_tokens", 128)
        temperature = data.get("temperature", 0.7)
        top_p = data.get("top_p", 0.9)
        top_k = data.get("top_k", 40)
        stream = data.get("stream", False)
        stop = data.get("stop", [])

        if isinstance(stop, str):
            stop = [stop]

        request_id = f"cmpl-{int(time.time() * 1000)}"

        if stream:
            self.send_response(200)
            self.send_header("Content-Type", "text/event-stream")
            self.send_header("Cache-Control", "no-cache")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()

            for chunk in self.model.generate_stream(prompt, max_tokens, temperature, top_k, top_p, stop):
                event = {
                    "id": request_id,
                    "object": "text_completion",
                    "created": int(time.time()),
                    "model": self.model.model_name,
                    "choices": [{
                        "text": chunk["text"],
                        "index": 0,
                        "logprobs": None,
                        "finish_reason": None,
                    }]
                }
                self.send_sse(event)

            # Send done event
            self.send_sse("[DONE]")

        else:
            text = self.model.generate(prompt, max_tokens, temperature, top_k, top_p, stop)
            self.send_json({
                "id": request_id,
                "object": "text_completion",
                "created": int(time.time()),
                "model": self.model.model_name,
                "choices": [{
                    "text": text,
                    "index": 0,
                    "logprobs": None,
                    "finish_reason": "stop",
                }],
                "usage": {
                    "prompt_tokens": len(self.model.encode(prompt)),
                    "completion_tokens": len(self.model.encode(text)),
                    "total_tokens": len(self.model.encode(prompt)) + len(self.model.encode(text)),
                }
            })

    def handle_chat_completions(self, data):
        """Handle /v1/chat/completions endpoint."""
        messages = data.get("messages", [])
        max_tokens = data.get("max_tokens", 128)
        temperature = data.get("temperature", 0.7)
        top_p = data.get("top_p", 0.9)
        top_k = data.get("top_k", 40)
        stream = data.get("stream", False)
        stop = data.get("stop", [])

        if isinstance(stop, str):
            stop = [stop]

        # Build prompt from messages
        prompt = ""
        for msg in messages:
            role = msg.get("role", "user")
            content = msg.get("content", "")
            if role == "system":
                prompt += f"System: {content}\n"
            elif role == "user":
                prompt += f"User: {content}\n"
            elif role == "assistant":
                prompt += f"Assistant: {content}\n"
        prompt += "Assistant:"

        request_id = f"chatcmpl-{int(time.time() * 1000)}"

        if stream:
            self.send_response(200)
            self.send_header("Content-Type", "text/event-stream")
            self.send_header("Cache-Control", "no-cache")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()

            for chunk in self.model.generate_stream(prompt, max_tokens, temperature, top_k, top_p, stop):
                event = {
                    "id": request_id,
                    "object": "chat.completion.chunk",
                    "created": int(time.time()),
                    "model": self.model.model_name,
                    "choices": [{
                        "index": 0,
                        "delta": {"content": chunk["text"]},
                        "finish_reason": None,
                    }]
                }
                self.send_sse(event)

            self.send_sse("[DONE]")

        else:
            text = self.model.generate(prompt, max_tokens, temperature, top_k, top_p, stop)
            self.send_json({
                "id": request_id,
                "object": "chat.completion",
                "created": int(time.time()),
                "model": self.model.model_name,
                "choices": [{
                    "index": 0,
                    "message": {
                        "role": "assistant",
                        "content": text,
                    },
                    "finish_reason": "stop",
                }],
                "usage": {
                    "prompt_tokens": len(self.model.encode(prompt)),
                    "completion_tokens": len(self.model.encode(text)),
                    "total_tokens": len(self.model.encode(prompt)) + len(self.model.encode(text)),
                }
            })


def main():
    parser = argparse.ArgumentParser(
        description="C-Kernel-Engine Streaming Inference Server",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument(
        "--model-dir",
        default=os.path.join(os.path.expanduser("~"), ".cache", "huggingface", "hub", "SmolLM-135M"),
        help="HuggingFace model directory"
    )
    parser.add_argument(
        "--build-dir",
        default="build/server",
        help="Build output directory"
    )
    parser.add_argument(
        "--host",
        default="0.0.0.0",
        help="Host to bind to"
    )
    parser.add_argument(
        "--port",
        type=int,
        default=8080,
        help="Port to listen on"
    )
    parser.add_argument(
        "--context",
        type=int,
        default=512,
        help="Context window size"
    )
    parser.add_argument(
        "--no-banner",
        action="store_true",
        help="Don't show startup banner"
    )
    args = parser.parse_args()

    if not args.no_banner:
        print(BANNER.format(version=VERSION))

    # Check model exists
    if not os.path.exists(os.path.join(args.model_dir, "config.json")):
        print(f"Error: Model not found at {args.model_dir}")
        print("Download with: python scripts/download_smollm.py --outdir " + args.model_dir)
        return 1

    print(f"Loading model from {args.model_dir}...")

    try:
        model = CKernelModel(
            model_dir=args.model_dir,
            build_dir=args.build_dir,
            context_len=args.context
        )
    except Exception as e:
        print(f"Error loading model: {e}")
        return 1

    # Set up handler
    CKernelHandler.model = model
    CKernelHandler.server_start_time = time.time()

    # Start server
    server = HTTPServer((args.host, args.port), CKernelHandler)
    print(f"\nServer running at http://{args.host}:{args.port}")
    print(f"API endpoints:")
    print(f"  POST /v1/completions      - Text completion")
    print(f"  POST /v1/chat/completions - Chat completion")
    print(f"  GET  /v1/models           - List models")
    print(f"  GET  /health              - Health check")
    print(f"\nPress Ctrl+C to stop.\n")

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down...")
        server.shutdown()

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/ck_server.c
/*
 * C-Kernel-Engine HTTP Streaming Server
 *
 * A simple HTTP server for streaming token generation.
 * Supports Server-Sent Events (SSE) for real-time streaming.
 *
 * API Endpoints:
 *   POST /v1/completions - Text completion with streaming
 *   GET  /health         - Health check
 *   GET  /               - Welcome message
 *
 * Build:
 *   make tools/ck_server
 *
 * Usage:
 *   ./build/ck_server --port 8080 --model weights.bin
 *
 * By Anthony Shivakumar
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include <unistd.h>
#include <signal.h>
#include <time.h>
#include <errno.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <pthread.h>
#include <sys/select.h>
#include <fcntl.h>

#define CK_VERSION "0.1.0"
#define MAX_REQUEST_SIZE (1024 * 1024)
#define MAX_RESPONSE_SIZE (1024 * 1024)
#define MAX_CONNECTIONS 64

static volatile sig_atomic_t g_shutdown = 0;

static void signal_handler(int sig) {
    (void)sig;
    g_shutdown = 1;
}

/* HTTP Response helpers */
static void send_response(int fd, int status, const char *content_type, const char *body) {
    const char *status_text = "OK";
    if (status == 400) status_text = "Bad Request";
    if (status == 404) status_text = "Not Found";
    if (status == 500) status_text = "Internal Server Error";

    char header[1024];
    int header_len = snprintf(header, sizeof(header),
        "HTTP/1.1 %d %s\r\n"
        "Content-Type: %s\r\n"
        "Content-Length: %zu\r\n"
        "Access-Control-Allow-Origin: *\r\n"
        "Connection: close\r\n"
        "\r\n",
        status, status_text, content_type, strlen(body));

    write(fd, header, header_len);
    write(fd, body, strlen(body));
}

static void send_json(int fd, int status, const char *json) {
    send_response(fd, status, "application/json", json);
}

static void start_sse(int fd) {
    const char *header =
        "HTTP/1.1 200 OK\r\n"
        "Content-Type: text/event-stream\r\n"
        "Cache-Control: no-cache\r\n"
        "Access-Control-Allow-Origin: *\r\n"
        "Connection: keep-alive\r\n"
        "\r\n";
    write(fd, header, strlen(header));
}

static void send_sse_event(int fd, const char *data) {
    char buf[4096];
    int len = snprintf(buf, sizeof(buf), "data: %s\n\n", data);
    write(fd, buf, len);
}

/* Simple JSON parsing helpers */
static const char *json_find_key(const char *json, const char *key) {
    char pattern[256];
    snprintf(pattern, sizeof(pattern), "\"%s\"", key);
    const char *pos = strstr(json, pattern);
    if (!pos) return NULL;
    pos = strchr(pos, ':');
    if (!pos) return NULL;
    pos++;
    while (*pos == ' ' || *pos == '\t') pos++;
    return pos;
}

static int json_parse_string(const char *pos, char *buf, int max_len) {
    if (*pos != '"') return -1;
    pos++;
    int len = 0;
    while (*pos && *pos != '"' && len < max_len - 1) {
        if (*pos == '\\' && pos[1]) {
            pos++;
            switch (*pos) {
                case 'n': buf[len++] = '\n'; break;
                case 'r': buf[len++] = '\r'; break;
                case 't': buf[len++] = '\t'; break;
                default: buf[len++] = *pos; break;
            }
        } else {
            buf[len++] = *pos;
        }
        pos++;
    }
    buf[len] = '\0';
    return len;
}

static int json_parse_int(const char *pos) {
    return atoi(pos);
}

static double json_parse_float(const char *pos) {
    return atof(pos);
}

static bool json_parse_bool(const char *pos) {
    return strncmp(pos, "true", 4) == 0;
}

/* Request handler context */
typedef struct {
    int client_fd;
    char *request;
    int request_len;
} RequestContext;

/* Demo token generator (placeholder) */
static const char *demo_tokens[] = {
    "Once", " upon", " a", " time", ",", " in", " a", " land",
    " far", " away", ",", " there", " lived", " a", " brave",
    " knight", ".", " He", " was", " known", " for", " his",
    " courage", " and", " wisdom", "."
};
static int demo_token_count = sizeof(demo_tokens) / sizeof(demo_tokens[0]);

static void handle_completions(int fd, const char *body, bool stream) {
    /* Parse request */
    char prompt[4096] = "";
    int max_tokens = 128;
    float temperature = 0.7f;

    const char *pos;
    if ((pos = json_find_key(body, "prompt"))) {
        json_parse_string(pos, prompt, sizeof(prompt));
    }
    if ((pos = json_find_key(body, "max_tokens"))) {
        max_tokens = json_parse_int(pos);
    }
    if ((pos = json_find_key(body, "temperature"))) {
        temperature = json_parse_float(pos);
    }
    if ((pos = json_find_key(body, "stream"))) {
        stream = json_parse_bool(pos);
    }

    printf("[INFO] Completion request: prompt='%.50s...' max_tokens=%d stream=%d\n",
           prompt, max_tokens, stream);

    if (stream) {
        start_sse(fd);

        /* Stream tokens */
        for (int i = 0; i < max_tokens && i < demo_token_count; i++) {
            char event[1024];
            snprintf(event, sizeof(event),
                "{\"choices\":[{\"text\":\"%s\",\"index\":0}],\"model\":\"ck-demo\"}",
                demo_tokens[i]);
            send_sse_event(fd, event);

            /* Simulate generation delay */
            usleep(50000);  /* 50ms per token */
        }

        send_sse_event(fd, "[DONE]");
    } else {
        /* Non-streaming: collect all tokens */
        char response[MAX_RESPONSE_SIZE];
        char text[8192] = "";
        int text_len = 0;

        for (int i = 0; i < max_tokens && i < demo_token_count; i++) {
            int tok_len = strlen(demo_tokens[i]);
            if (text_len + tok_len < (int)sizeof(text) - 1) {
                strcpy(text + text_len, demo_tokens[i]);
                text_len += tok_len;
            }
        }

        snprintf(response, sizeof(response),
            "{\"choices\":[{\"text\":\"%s\",\"index\":0,\"finish_reason\":\"length\"}],"
            "\"model\":\"ck-demo\",\"usage\":{\"prompt_tokens\":%d,\"completion_tokens\":%d}}",
            text, (int)strlen(prompt) / 4, max_tokens);

        send_json(fd, 200, response);
    }
}

static void handle_request(int client_fd, const char *request, int request_len) {
    /* Parse request line */
    char method[16] = "";
    char path[256] = "";

    sscanf(request, "%15s %255s", method, path);

    printf("[INFO] %s %s\n", method, path);

    /* Find request body */
    const char *body = strstr(request, "\r\n\r\n");
    if (body) body += 4;

    /* Route request */
    if (strcmp(method, "OPTIONS") == 0) {
        /* CORS preflight */
        const char *response =
            "HTTP/1.1 200 OK\r\n"
            "Access-Control-Allow-Origin: *\r\n"
            "Access-Control-Allow-Methods: GET, POST, OPTIONS\r\n"
            "Access-Control-Allow-Headers: Content-Type\r\n"
            "Content-Length: 0\r\n"
            "\r\n";
        write(client_fd, response, strlen(response));
    }
    else if (strcmp(method, "GET") == 0 && strcmp(path, "/") == 0) {
        /* Welcome page */
        char html[2048];
        snprintf(html, sizeof(html),
            "<!DOCTYPE html>\n"
            "<html><head><title>C-Kernel-Engine Server</title></head>\n"
            "<body>\n"
            "<h1>C-Kernel-Engine Inference Server v%s</h1>\n"
            "<p>By Anthony Shivakumar</p>\n"
            "<h2>API Endpoints:</h2>\n"
            "<ul>\n"
            "<li>POST /v1/completions - Text completion</li>\n"
            "<li>GET /health - Health check</li>\n"
            "</ul>\n"
            "<h2>Example:</h2>\n"
            "<pre>\n"
            "curl -X POST http://localhost:8080/v1/completions \\\n"
            "  -H 'Content-Type: application/json' \\\n"
            "  -d '{\"prompt\": \"Hello\", \"max_tokens\": 50, \"stream\": true}'\n"
            "</pre>\n"
            "</body></html>\n",
            CK_VERSION);
        send_response(client_fd, 200, "text/html", html);
    }
    else if (strcmp(method, "GET") == 0 && strcmp(path, "/health") == 0) {
        char json[256];
        snprintf(json, sizeof(json),
            "{\"status\":\"ok\",\"version\":\"%s\"}", CK_VERSION);
        send_json(client_fd, 200, json);
    }
    else if (strcmp(method, "GET") == 0 && strcmp(path, "/v1/models") == 0) {
        send_json(client_fd, 200,
            "{\"data\":[{\"id\":\"ck-demo\",\"object\":\"model\",\"owned_by\":\"c-kernel-engine\"}]}");
    }
    else if (strcmp(method, "POST") == 0 &&
             (strcmp(path, "/v1/completions") == 0 ||
              strcmp(path, "/v1/chat/completions") == 0)) {
        if (body) {
            handle_completions(client_fd, body, false);
        } else {
            send_json(client_fd, 400, "{\"error\":\"No request body\"}");
        }
    }
    else {
        send_json(client_fd, 404, "{\"error\":\"Not found\"}");
    }
}

static void *client_thread(void *arg) {
    RequestContext *ctx = (RequestContext *)arg;

    handle_request(ctx->client_fd, ctx->request, ctx->request_len);

    close(ctx->client_fd);
    free(ctx->request);
    free(ctx);
    return NULL;
}

static void print_banner(void) {
    printf("\n");
    printf("\033[36m\033[1m");
    printf("   ____      _  __                    _   ____                           \n");
    printf("  / ___|    | |/ /___ _ __ _ __   ___| | / ___|  ___ _ ____   _____ _ __ \n");
    printf(" | |   _____| ' // _ \\ '__| '_ \\ / _ \\ | \\___ \\ / _ \\ '__\\ \\ / / _ \\ '__|\n");
    printf(" | |__|_____| . \\  __/ |  | | | |  __/ |  ___) |  __/ |   \\ V /  __/ |   \n");
    printf("  \\____|    |_|\\_\\___|_|  |_| |_|\\___|_| |____/ \\___|_|    \\_/ \\___|_|   \n");
    printf("\033[0m");
    printf("\n");
    printf("\033[32m  C-Kernel-Engine Server v%s\033[0m\n", CK_VERSION);
    printf("  By Anthony Shivakumar\n");
    printf("\n");
    printf("  Streaming Inference API Server\n");
    printf("  ============================================================\n");
    printf("\n");
}

int main(int argc, char **argv) {
    int port = 8080;
    const char *host = "0.0.0.0";
    const char *model_path = NULL;
    bool show_banner = true;

    /* Parse arguments */
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--port") == 0 && i + 1 < argc) {
            port = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--host") == 0 && i + 1 < argc) {
            host = argv[++i];
        } else if (strcmp(argv[i], "--model") == 0 && i + 1 < argc) {
            model_path = argv[++i];
        } else if (strcmp(argv[i], "--no-banner") == 0) {
            show_banner = false;
        } else if (strcmp(argv[i], "--help") == 0) {
            printf("Usage: %s [options]\n\n", argv[0]);
            printf("Options:\n");
            printf("  --port N      Port to listen on (default: 8080)\n");
            printf("  --host ADDR   Host to bind to (default: 0.0.0.0)\n");
            printf("  --model PATH  Path to model weights\n");
            printf("  --no-banner   Don't show startup banner\n");
            printf("  --help        Show this help\n");
            return 0;
        }
    }

    if (show_banner) {
        print_banner();
    }

    /* Set up signal handler */
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    signal(SIGPIPE, SIG_IGN);

    if (model_path) {
        printf("Loading model from: %s\n", model_path);
        /* TODO: Load actual model */
    } else {
        printf("\033[33mWarning: No model loaded, using demo responses.\033[0m\n");
        printf("Use --model PATH to load actual weights.\n\n");
    }

    /* Create socket */
    int server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd < 0) {
        perror("socket");
        return 1;
    }

    int opt = 1;
    setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    struct sockaddr_in addr = {0};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(port);
    addr.sin_addr.s_addr = inet_addr(host);

    if (bind(server_fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("bind");
        close(server_fd);
        return 1;
    }

    if (listen(server_fd, MAX_CONNECTIONS) < 0) {
        perror("listen");
        close(server_fd);
        return 1;
    }

    printf("Server listening on http://%s:%d\n", host, port);
    printf("\nEndpoints:\n");
    printf("  POST /v1/completions      - Text completion\n");
    printf("  POST /v1/chat/completions - Chat completion\n");
    printf("  GET  /v1/models           - List models\n");
    printf("  GET  /health              - Health check\n");
    printf("\nPress Ctrl+C to stop.\n\n");

    /* Accept loop with select() for interruptible waiting */
    while (!g_shutdown) {
        fd_set read_fds;
        struct timeval tv;

        FD_ZERO(&read_fds);
        FD_SET(server_fd, &read_fds);

        /* Wait up to 1 second for a connection */
        tv.tv_sec = 1;
        tv.tv_usec = 0;

        int sel = select(server_fd + 1, &read_fds, NULL, NULL, &tv);
        if (sel < 0) {
            if (errno == EINTR) continue;  /* Signal interrupted, check g_shutdown */
            perror("select");
            break;
        }
        if (sel == 0) {
            /* Timeout - loop back and check g_shutdown */
            continue;
        }

        struct sockaddr_in client_addr;
        socklen_t client_len = sizeof(client_addr);

        int client_fd = accept(server_fd, (struct sockaddr *)&client_addr, &client_len);
        if (client_fd < 0) {
            if (errno == EINTR) continue;
            perror("accept");
            continue;
        }

        /* Read request */
        char *request = (char *)malloc(MAX_REQUEST_SIZE);
        if (!request) {
            close(client_fd);
            continue;
        }

        int request_len = read(client_fd, request, MAX_REQUEST_SIZE - 1);
        if (request_len <= 0) {
            free(request);
            close(client_fd);
            continue;
        }
        request[request_len] = '\0';

        /* Handle in thread */
        RequestContext *ctx = (RequestContext *)malloc(sizeof(RequestContext));
        ctx->client_fd = client_fd;
        ctx->request = request;
        ctx->request_len = request_len;

        pthread_t thread;
        if (pthread_create(&thread, NULL, client_thread, ctx) != 0) {
            handle_request(client_fd, request, request_len);
            close(client_fd);
            free(request);
            free(ctx);
        } else {
            pthread_detach(thread);
        }
    }

    printf("\nShutting down...\n");
    close(server_fd);
    return 0;
}
